Miron Banjac - Abh Data Analytics Internship

Currently on task 1 writing modulated code in for each analyser to be used.

In [2]:
import numpy as np
import pandas as pd
#import requests
from urllib.request import urlopen
import json

chicago_api = urlopen('https://data.cityofchicago.org/resource/uupf-x98q.json')
chicago_json = json.loads(chicago_api.read())
#after json.loads can use pandas.DataFrame.from_dict() to convert JSON to DataFrame
#print(pd.json_normalize(chicago_json))
data_limited = pd.read_json('https://data.cityofchicago.org/resource/uupf-x98q.json')
data = pd.read_csv('Business_Licenses_-_Current_Active.csv', low_memory=False)
#data = pd.json_normalize(chicago_json)
print(data.shape)

data.columns

(57369, 34)


Index(['ID', 'LICENSE ID', 'ACCOUNT NUMBER', 'SITE NUMBER', 'LEGAL NAME',
       'DOING BUSINESS AS NAME', 'ADDRESS', 'CITY', 'STATE', 'ZIP CODE',
       'WARD', 'PRECINCT', 'WARD PRECINCT', 'POLICE DISTRICT', 'LICENSE CODE',
       'LICENSE DESCRIPTION', 'BUSINESS ACTIVITY ID', 'BUSINESS ACTIVITY',
       'LICENSE NUMBER', 'APPLICATION TYPE', 'APPLICATION CREATED DATE',
       'APPLICATION REQUIREMENTS COMPLETE', 'PAYMENT DATE',
       'CONDITIONAL APPROVAL', 'LICENSE TERM START DATE',
       'LICENSE TERM EXPIRATION DATE', 'LICENSE APPROVED FOR ISSUANCE',
       'DATE ISSUED', 'LICENSE STATUS', 'LICENSE STATUS CHANGE DATE', 'SSA',
       'LATITUDE', 'LONGITUDE', 'LOCATION'],
      dtype='object')

In [3]:
#making first wrapper function for task 1 analysis

from statistics import mean

def function_property_of_column(func, column_name):
    return func(data[column_name])

print(function_property_of_column(mean, 'SITE NUMBER'))
print(mean(data['SITE NUMBER']))

6.277327476511705
6.277327476511705


In [4]:
#first analysis: missing values and types

#print(data.head())
data.info()
#Alternatively can use just .count() to get the number of non-missing values for each column
#print(data.count())


data.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57369 entries, 0 to 57368
Data columns (total 34 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   ID                                 57369 non-null  object 
 1   LICENSE ID                         57369 non-null  int64  
 2   ACCOUNT NUMBER                     57369 non-null  int64  
 3   SITE NUMBER                        57369 non-null  int64  
 4   LEGAL NAME                         57369 non-null  object 
 5   DOING BUSINESS AS NAME             57290 non-null  object 
 6   ADDRESS                            57369 non-null  object 
 7   CITY                               57369 non-null  object 
 8   STATE                              57369 non-null  object 
 9   ZIP CODE                           57359 non-null  object 
 10  WARD                               52958 non-null  float64
 11  PRECINCT                           40258 non-null  flo

ID                                    object
LICENSE ID                             int64
ACCOUNT NUMBER                         int64
SITE NUMBER                            int64
LEGAL NAME                            object
DOING BUSINESS AS NAME                object
ADDRESS                               object
CITY                                  object
STATE                                 object
ZIP CODE                              object
WARD                                 float64
PRECINCT                             float64
WARD PRECINCT                         object
POLICE DISTRICT                      float64
LICENSE CODE                           int64
LICENSE DESCRIPTION                   object
BUSINESS ACTIVITY ID                  object
BUSINESS ACTIVITY                     object
LICENSE NUMBER                         int64
APPLICATION TYPE                      object
APPLICATION CREATED DATE              object
APPLICATION REQUIREMENTS COMPLETE     object
PAYMENT DA

In [5]:
#go through each column in pandas dataframe and, for object type, describe them, for int and float types, give a summary.

for column in data:
    print(data[column].describe())

#print(data['LONGITUDE'].describe())
#print(data['LATITUDE'].describe())

count                57369
unique               57368
top       2823191-20211208
freq                     2
Name: ID, dtype: object
count    5.736900e+04
mean     2.786946e+06
std      3.632965e+04
min      2.102725e+06
25%      2.757683e+06
50%      2.787988e+06
75%      2.816571e+06
max      2.851375e+06
Name: LICENSE ID, dtype: float64
count     57369.000000
mean     307456.862068
std      164575.002892
min           1.000000
25%      214704.000000
50%      372433.000000
75%      430228.000000
max      486617.000000
Name: ACCOUNT NUMBER, dtype: float64
count    57369.000000
mean         6.277327
std         29.002589
min          1.000000
25%          1.000000
50%          1.000000
75%          2.000000
max        589.000000
Name: SITE NUMBER, dtype: float64
count            57369
unique           39460
top       WALGREEN CO.
freq               204
Name: LEGAL NAME, dtype: object
count      57290
unique     41940
top       SUBWAY
freq         123
Name: DOING BUSINESS AS NAME, dtype:

In [6]:
#frequency of different observations for each column in data

for column in data:
    print(data[column].value_counts())

2823191-20211208    2
2850614-20220527    1
1622615-20210916    1
1446227-20210916    1
2797610-20210831    1
                   ..
2073612-20210116    1
1880003-20220116    1
2130243-20220116    1
2767249-20210115    1
2189662-20221016    1
Name: ID, Length: 57368, dtype: int64
2850614    1
2800243    1
2797610    1
2808610    1
2798402    1
          ..
2820970    1
2821237    1
2767249    1
2762084    1
2839926    1
Name: LICENSE ID, Length: 57369, dtype: int64
147       204
63160     197
7551      141
63414     123
11071     107
         ... 
313895      1
341537      1
38820       1
455401      1
348878      1
Name: ACCOUNT NUMBER, Length: 39586, dtype: int64
1      38172
2       7906
3       3285
4       1557
5        873
       ...  
264        1
485        1
510        1
479        1
540        1
Name: SITE NUMBER, Length: 372, dtype: int64
WALGREEN CO.                          204
SP PLUS CORPORATION                   197
BOND DRUG COMPANY OF ILLINOIS, LLC    141
STARBUCKS COR

In [7]:
#seperating the data by type

int_data = data.select_dtypes('int')
float_data = data.select_dtypes('float')
object_data = data.select_dtypes('object')

print(int_data.shape)
print(float_data.shape)
print(object_data.shape)

(57369, 5)
(57369, 7)
(57369, 22)


In [8]:
#converting data into correct type
#data['APPLICATION CREATED DATE'] = data['APPLICATION CREATED DATE'].astype(str)
data['APPLICATION CREATED DATE NEW'] = pd.to_datetime(data['APPLICATION CREATED DATE'], format='%m/%d/%Y', errors = 'coerce')

In [15]:
#data['DOING BUSINESS AS NAME'].astype(str).describe()
data['SSA'].value_counts()


33.0    1183
10.0     963
17.0     665
31.0     588
3.0      565
1.0      561
27.0     547
8.0      536
34.0     533
60.0     514
29.0     498
25.0     466
38.0     444
43.0     389
51.0     383
50.0     354
21.0     320
39.0     319
61.0     317
26.0     315
13.0     311
22.0     285
23.0     285
24.0     271
35.0     254
42.0     225
59.0     216
20.0     197
18.0     176
2.0      157
14.0     146
49.0     145
47.0     139
48.0     137
69.0     137
28.0     137
32.0     131
5.0      130
45.0     113
54.0     106
4.0      100
63.0      99
40.0      98
56.0      91
55.0      90
19.0      90
16.0      86
41.0      64
52.0      47
44.0      36
62.0      36
7.0       23
64.0       8
Name: SSA, dtype: int64

Chicago Buisness list Task 1 analysis.

I am doing the appropriate analyzers for each of the relevant variables within the data frame.

ID:
Type: text
Unique Entries: 57368
Missing: 0
Except for two tha match under the id:2823191-20211208. Upon lookup in the table in sql the two observaions are identical except LICENSE ID, BUSINESS ACTIVITY ID and PAYMENT DATE was null in one of them. This could be a misteak or a double entry.

LICENSE ID:
Type: integer
Unique: 57369 (all)
Missing: 0
Smallest id:2102725, Largest: 2851375
Each license has a unique LICENSE ID assigned to it.

ACCOUNT NUMBER:
Type: integer
Unique: 39586
Missing: 0
This is the account number of the owner of the business which explains why there are repeats within this column as multiple licenses under the owner.

SITE NUMBER:
Type: integer
Unique: 372
Missing: 0
Range: 1 to 540
The 5 zones with the most business license:
1      38172
2       7906
3       3285
4       1557
5        873

LEGAL NAME:
Type: text
Unique: 39460
Missing: 0





